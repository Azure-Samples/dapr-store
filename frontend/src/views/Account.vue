<template>
  <div>
    <h1>
      User Account
      <b-button size="lg" variant="danger" class="float-right" @click="logout">
        <fa icon="sign-out-alt" /> &nbsp; LOGOUT
      </b-button>
    </h1>

    <table v-if="user.account && user.userName" class="table">
      <tr><td>Display Name</td><td>{{ user.account.name }}</td></tr>
      <tr><td>Username</td><td>{{ user.userName }}</td></tr>
      <tr><td>Unique ID</td><td>{{ user.account.accountIdentifier }}</td></tr>
    </table>

    <div />
  </div>
</template>

<script>
import { userProfile } from '../main'
import User from '../user'

export default {
  name: 'Account',

  data() {
    return {
      user: userProfile
    }
  },

  created() {
    if (!userProfile.token) {
      this.$router.replace({ path: '/' })
    }
  },

  methods: {
    async logout() {
      Object.assign(userProfile, new User())
      localStorage.removeItem('user')
      localStorage.removeItem('cart')
      //await msalApp.logout()

      this.$router.push({ name: 'home' })
    }
  }
}
</script>

<style scoped>
.table {
  font-size: 140%;
}
</style>