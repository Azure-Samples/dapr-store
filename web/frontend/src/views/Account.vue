<!--
// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - User / account details view
// ----------------------------------------------------------------------------
-->

<template>
  <div>
    <error-box :error="error" />
    <h1>
      User Account
      <b-button size="lg" variant="danger" class="float-right" @click="logout">
        <fa icon="sign-out-alt" /> &nbsp; LOGOUT
      </b-button>
    </h1>
    <br>

    <div v-if="!registeredUser" class="text-center">
      <b-spinner variant="success" style="width: 5rem; height: 5rem;" />
    </div>
    <b-card v-if="registeredUser" class="details">
      <img class="profile d-none d-md-block" :src="photo">

      Display Name: <b>{{ registeredUser.displayName }}</b>
      <br>
      Username: <b>{{ registeredUser.username }}</b>
    </b-card>

    <br>
    <h1>
      Orders
      <b-button size="lg" variant="success" class="float-right" @click="reloadOrders">
        <fa icon="redo-alt" /> &nbsp; Refresh
      </b-button>
    </h1>

    <div v-if="!ordersLoaded" class="text-center">
      <b-spinner variant="success" style="width: 5rem; height: 5rem;" />
    </div>

    <order v-for="order in orders" :key="order.id" :order="order" />
  </div>
</template>

<script>
import api from '../services/api'
import auth from '../services/auth'
import graph from '../services/graph'
import ErrorBox from '../components/ErrorBox'
import Order from '../components/Order'

export default {
  name: 'Account',

  components: {
    'error-box': ErrorBox,
    'order': Order
  },

  data() {
    return {
      registeredUser: auth.user(),
      photo: 'img/placeholder-profile.jpg',
      error: null,
      orders: null,
      ordersLoaded: false
    }
  },

  async created() {
    try {
      if (auth.user()) {
        let resp = await api.userGet(auth.user().userName)
        if (resp) {
          this.registeredUser = resp
        }
        graph.getPhoto()
          .then((photo) => { if (photo){ this.photo = photo } })
      }
    } catch (err) {
      this.error = err
    }

    this.reloadOrders()
  },

  methods: {
    async logout() {
      await auth.logout()

      this.$router.push({ name: 'home' })
    },

    async reloadOrders() {
      this.ordersLoaded = false
      this.orders = []
      let orderList = []

      try {
        orderList = await api.ordersForUser(auth.user().userName)
      } catch (err) {
        this.error = err
      }

      // If you have no orders, skip it
      if (!orderList) {
        this.ordersLoaded = true
        return
      }

      // Load orders call the API to fetch details
      for (let orderId of orderList.reverse()) {
        try {
          let order = await api.orderGet(orderId)
          this.orders.push(order)
        } catch (err) {
          this.error += err
          continue
        }
      }

      this.ordersLoaded = true
    }
  }
}
</script>

<style scoped>
  code {
    color:rgb(23, 38, 173);
    font-size: 1.2rem;
  }

  .details {
    font-size: 140%;
  }

  .profile {
    float: right;
    width: 10rem;
    border-radius: 50%;
  }
</style>