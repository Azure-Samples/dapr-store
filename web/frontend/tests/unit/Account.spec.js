import { createLocalVue, mount } from '@vue/test-utils'
import flushPromises from 'flush-promises'

import BootstrapVue from 'bootstrap-vue'
const localVue = createLocalVue()
localVue.use(BootstrapVue)

import VueRouter from 'vue-router'
localVue.use(VueRouter)
const router = new VueRouter()

import Account from '@/views/Account.vue'

jest.mock('@/services/api')
jest.mock('@/services/auth')

describe('Account.vue', () => {
  it('renders user profile', async () => {
    const wrapper = mount(Account, {
      localVue,
      router
    })

    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })
})