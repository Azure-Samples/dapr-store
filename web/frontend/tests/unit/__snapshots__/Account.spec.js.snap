// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account.vue renders user profile 1`] = `
<div>
  <!---->
  <h1>
    User Account
    <button type="button" class="btn float-right btn-danger btn-lg">
      <fa icon="sign-out-alt"></fa> &nbsp; LOGOUT
    </button></h1> <br>
  <!---->
  <div class="card details">
    <!---->
    <!---->
    <div class="card-body">
      <!---->
      <!----><img src="img/placeholder-profile.jpg" class="profile d-none d-md-block">

      Display Name: <b>Mock User</b> <br>
      Username: <b>mock@example.net</b></div>
    <!---->
    <!---->
  </div> <br>
  <h1>
    Orders
    <button type="button" class="btn float-right btn-success btn-lg">
      <fa icon="redo-alt"></fa> &nbsp; Refresh
    </button></h1>
  <!---->
  <div class="card order">
    <!---->
    <!---->
    <div class="card-body">
      <!---->
      <!----><a href="#/order/ord-mock" class="" target="_self">
        <h2>Mock order</h2>
      </a>
      <h2>
        Status: <span class="text-capitalize order-status order-new">new</span></h2>
      <ul>
        <li>
          Order Id: ord-mock
        </li>
        <li>Amount: £22.4</li>
        <li>Items:</li>
        <ul>
          <li>
            2 x Paisley Cravat Ascot Tie — £11.2
          </li>
        </ul>
      </ul>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
